walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,180 @@
+      LinkedList<TimeRange> time, LinkedList<String> attendees, Event event) {
+    if (!Collections.disjoint(attendees, event.getAttendees())) {
+      // Go through all the free time slots and check from which one is the event overlapping
+      ListIterator<TimeRange> iterator = time.listIterator();
+      while (iterator.hasNext()) {
+        TimeRange freeSlot = iterator.next();
+        // When th eoverlapping free slot is found send our free time and event slot to get altered.
+        if (freeSlot.overlaps(event.getWhen())) {
+          updateFreeTime(freeSlot, event.getWhen(), time, iterator);
+          // After updating free time is done, check if the event ends in this free slots
+          if (event.getWhen().end() <= freeSlot.end()) break;
+        }
+      }
+    }
+  }
+
+  /**
+   * This method free time, free time slot and current event with our free time slot and the iterator
+   * and checks all conditions in which the event and time slot could overlap. 
+   * Then calculates the free time minus overlapping part. 
+   * It checks the new slot with the meeting slot and then adds or removes it. 
+   * (By removing/ setting or adding in the iterator).
+   */
+  private void updateFreeTime(
+      TimeRange freeSlot,
+      TimeRange eventSlot,
+      LinkedList<TimeRange> freeTime,
+      ListIterator<TimeRange> iterator) {
+
+    int freeSlotStart = freeSlot.start();
+    int freeSlotEnd = freeSlot.end();
+
+    int eventSlotStart = eventSlot.start();
+    int eventSlotEnd = eventSlot.end();
+
+    TimeRange newSlot;
+
+    // Case: Free time: |---|
+    // Event slot:    |-------|
+    if (eventSlotStart <= freeSlotStart && eventSlotEnd >= freeSlotEnd) {
+      iterator.remove();
+    }
+
+    // Case: Free time: |--------|
+    // Event slot:     |-----|
+    // or
+    // |------|
+    // |--|
+    else if (eventSlotStart <= freeSlotStart && eventSlotEnd < freeSlotEnd) {
+      newSlot = TimeRange.fromStartEnd(eventSlotEnd, freeSlotEnd, false);
+      if (isSlotTimeEnough(newSlot)) iterator.set(newSlot);
+      else iterator.remove();
+    }
+
+    // Case: Free time: |--------|
+    // Event slot:            |-----|
+    // or
+    // |------|
+    //     |--|
+    else if (eventSlotStart > freeSlotStart && eventSlotEnd >= freeSlotEnd) {
+      newSlot = TimeRange.fromStartEnd(freeSlotStart, eventSlotStart, false);
+      if (isSlotTimeEnough(newSlot)) iterator.set(newSlot);
+      else iterator.remove();
+
+    }
+
+    // Case: Free time: |---------|
+    // Event slot:        |-----|
+    else if (eventSlotStart > freeSlotStart && eventSlotEnd < freeSlotEnd) {
+      newSlot = TimeRange.fromStartEnd(freeSlotStart, eventSlotStart, false);
+      if (isSlotTimeEnough(newSlot)) iterator.set(newSlot);
+      else iterator.remove();
+
+      newSlot = TimeRange.fromStartEnd(eventSlotEnd, freeSlotEnd, false);
+
+      if (isSlotTimeEnough(newSlot)) iterator.add(newSlot);
+      else iterator.remove();
```

Comment:
If you are removing the iterator in the first if statement, I don't think you can use it in the second if statement. In particular, you should not attempt to delete it twice.

I suggest you first do the slot after the event and don't remove the iterator and then do the slot before the event:
```
TimeRange afterSlot = TimeRange.fromStartEnd(eventSlotEnd, freeSlotEnd, false);
if (isSlotTimeEnough(afterSlot )) iterator.add(afterSlot);

TimeRange beforeSlot = TimeRange.fromStartEnd(freeSlotStart, eventSlotStart, false);
if (isSlotTimeEnough(beforeSlot )) iterator.set(beforeSlot);
else iterator.remove();
```

Reply:
Ohhh yes! 
I will do that!!!

Reply:
Umm..... it gave me a "java.lang.IllegalStateException" cause we are not doing iterator.next after adding stuff to it.
So I tried doing it with a boolean removedFirst flag.
Please let me know if that is a correct alternative to this problem.

Reply:
["java.lang.IllegalStateException" Stack overflow link](https://stackoverflow.com/questions/22361194/iterator-remove-illegalstateexception)

Reply:
Umm..... it gave me a ["java.lang.IllegalStateException"](https://stackoverflow.com/questions/22361194/iterator-remove-illegalstateexception) cause we are not doing iterator.next after adding stuff to it.
So I tried doing it with a boolean removedFirst flag.
Please let me know if that is a correct alternative to this problem.

Reply:
solution*